feat: construct pattern for deploying Langfuse Open Source

athewsey opened the pull request:
**Related issue:** #1031

**Description of changes:**

As discussed in the linked RFC, add a new L3 pattern for deployment of Langfuse (Open Source version) on ECS Fargate and AWS managed database/storage services. Code is originally derived from already work published under [aws-samples/deploy-langfuse-on-ecs-with-fargate](https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate), but significantly modified in addition to translation from Python to TypeScript and App to Construct set.


**⚠️ Important notes / caveats:**

- This change introduces a **new dependency** on [cdklabs/cdk-ecr-deployment](https://github.com/cdklabs/cdk-ecr-deployment) construct library, which we use for deploying the third-party Langfuse & ClickHouse container images into the user's ECR.
- While the constructs have been tested in usage, we have only the very beginnings of automated test coverage at the moment... Hoping to expand on that going forward.
    - I am seeing some test failures on local build, but they seem to be coming from packages we haven't touched? esp. `bedrock/data-sources`.


**Open questions / actions:**

There's probably a bit more work to do on this PR, but raising early to request feedback & give visibility!

In particular:

1. Should we promote ClickHouse to its own pattern?
    - [ClickHouse](https://clickhouse.com/), an Open Source OLAP RDBMS, is a required component for running Langfuse but I guess is an application users could potentially be interested in deploying in other contexts too? For now we've exposed it from our `index.ts`, but kept it within the Langfuse pattern folder.
2. ~Is the proposed construct ID `C8` available, or should be changed?~ - Using C11
3. Are we using the `BaseClass` correctly? Not sure I quite follow the purpose of listing out Lambda functions in `updateConstructUsageMetricCode`...
4. Still in progress: Parallel PR for https://github.com/aws-samples/generative-ai-cdk-constructs-samples to demonstrate usage in a deployable app.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

athewsey commented:
Whoops, realise I edited package.json directly instead of adding the dep via Projen - pushed a fix

scottschreckengaust commented on apidocs/enumerations/ConstructName.md:
```diff
@@ -40,3 +40,9 @@ Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+### LANGFUSEDEPLOYMENT
+
+> **LANGFUSEDEPLOYMENT**: `"C8"`
```
Checking on available values

scottschreckengaust replied:
```suggestion
> **LANGFUSEDEPLOYMENT**: `"C11"`
```

scottschreckengaust commented on src/common/base-class/construct-name-enum.ts:
```diff
@@ -16,4 +16,5 @@ export enum ConstructName {
   CUSTOMSAGEMAKERENDPOINT = 'C5',
   HUGGINGFACESAGEMAKERENDPOINT = 'C6',
   JUMPSTARTSAGEMAKERENDPOINT = 'C7',
+  LANGFUSEDEPLOYMENT = 'C8',  // TODO: Pending confirmation
```
```suggestion
  LANGFUSEDEPLOYMENT = 'C11',
```

athewsey replied:
Fixed in af673346a3a04d2cb330602ee58b1ae0807f609f

krokoko commented:
Hi @athewsey , thanks for this contribution ! I'm just back from ooo, will dive into it this week and put my comments here. Could you please add an architecture diagram for the construct ? Thank you !

athewsey commented:
Hi @krokoko thanks for taking a look. The diagram was actually in there already, but looks like the link from the README was broken. I've just fixed that, and rebased the branch against current main.

krokoko commented on .projenrc.ts:
```diff
@@ -75,6 +75,7 @@ const project = new awscdk.AwsCdkConstructLibrary({
   bundledDeps: [
     'deepmerge',
     `@aws-cdk/aws-lambda-python-alpha@${CDK_VERSION}-alpha.0`,
+    'cdk-ecr-deployment@3.0.154',
```
It seems versions before V4 are no longer supported: https://www.npmjs.com/package/cdk-ecr-deployment

krokoko commented on assets/clickhouse-container/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+ARG BASE_IMAGE=clickhouse:latest
```
image needs to be pulled from ECR

athewsey replied:
As far as I can tell ClickHouse don't seem to publish their images directly to ECR Public... The closest would be https://gallery.ecr.aws/bitnami/clickhouse that appears to be managed by a verified 3rd party Bitnami. Would it still be preferable to pull from ECR Public in that case?

(Do note the constructs also pull Langfuse image from Docker Hub - there don't seem to be any verified accounts hosting Langfuse versions on ECR Public)

krokoko commented:
Thanks @athewsey for raising this PR !

I think overall it is a great construct, however it seems pretty complex. Many moving parts and parameters. We should think about how to break it down into smaller, more manageable pieces as per our [design guidelines](https://github.com/awslabs/generative-ai-cdk-constructs/blob/main/DESIGN_GUIDELINES.md) 

Per experience, such large constructs often suffer from low adoption rates and impose significant cognitive load on users. Those constructs also require significant maintenance efforts, and provide limited flexibility and customization for end users.

Regarding the base class, it was heavily used in some constructs we deprecated recently, and we may not have use of it anymore (@dineshSajwan - checking with you to keep me honest here) so you might skip this part !
